pkg/store/tsdb.go

```diff
@@ -0,0 +1,181 @@
+package store
+
+import (
+	"context"
+	"math"
+	"sort"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/tsdb/chunks"
+
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/labels"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type TSDBStore struct {
+	logger log.Logger
+	db     *tsdb.DB
+	labels labels.Labels
+}
+
+// NewTSDBStore implements the store API against a local TSDB instance.
+// It attaches the provided external labels to all results.
+func NewTSDBStore(logger log.Logger, reg prometheus.Registerer, db *tsdb.DB, externalLabels labels.Labels) *TSDBStore {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	return &TSDBStore{
+		logger: logger,
+		db:     db,
+		labels: externalLabels,
+	}
+}
+
+// Info returns store information about the Prometheus instance.
+func (s *TSDBStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb.InfoResponse, error) {
+	res := &storepb.InfoResponse{
+		Labels: make([]storepb.Label, 0, len(s.labels)),
+	}
+	for _, l := range s.labels {
+		res.Labels = append(res.Labels, storepb.Label{
+			Name:  l.Name,
+			Value: l.Value,
+		})
+	}
+	return res, nil
+}
+
+// Series returns all series for a requested time range and label matcher. The returned data may
+// exceed the requested time bounds.
+//
+// Prometheus's range query API is not suitable to give us all datapoints. We use the
+// instant API and do a range selection in PromQL to cover the queried time range.
+func (s *TSDBStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	match, newMatchers, err := extLabelsMatches(s.labels, r.Matchers)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+	if !match {
+		return nil
+	}
+	matchers, err := translateMatchers(newMatchers)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	// TODO(fabxc): an improvement over this trivial approach would be to directly
```

Comment:
`s/an/An`

Reply:
Also: Why we need that improvement? What is the bottleneck here? (: Sorry.. it is not so straightforward.. We use chunks provided by TSDB right? We just encode it to match our gRPC proto

Reply:
Yes, we would need to transform to `storepb.Chunk` anyway. But the `.Data` field is just an opaque byte slice and we could just take those bytes as they exist in TSDB since the chunk XOR encoding is the same.

Instead we decode chunks in TSDB and create a new chunk. That's just wasted work really.
But it's easier for now and I don't know how performance relevant it would be anyway.